Add LargeDenseBlock Representations and support for Strings

kev-inn opened the pull request:
Adds a LargeDenseBlock representation for the DenseRowBlocks, while also fixing Bugs of the DenseRowBlocks.

Also adds a new DenseRowBlock for Strings (also a LDRB).

Tests were also added for LDRBs, while also adding some for not yet tested functionality.

Tests currently do not test LDRBs for large (greater than Integer.MAX_VALUE) array requirements.

kev-inn committed:
Add LargeDenseBlock Representations with new support for Strings for DRB

Adds a LargeDenseBlock representation for the DenseRowBlocks, while also fixing Bugs of the DenseRowBlocks.

Also adds a new DenseRowBlock for Strings (also a LDRB).

Tests were also added for LDRBs, while also adding some for not yet tested functionality.

Tests currently do not test LDRBs for large (greater than Integer.MAX_VALUE) array requirements.

mboehm7 committed:
[MINOR] Cleanup unnecessary mapreduce dependencies from pom.xml

mboehm7 committed:
[MINOR] Remove R dev/package installation from travis config

mboehm7 committed:
[SYSTEMDS-122] Reduced sparse buffer pool serialization overhead

This patch makes the shallow serialize threshold slightly more
aggressive, which means that we accept larger in-memory overhead in
order to avoid deep serialization of sparse matrices in the buffer pool.

On a parfor data augmentation application with 10K input images and 20K
output images, this patch improves performance from 9.2s (139s left
indexing, 46s buffer pool release) to 6.4s (82s left indexing, 7.9s
buffer pool release).

mboehm7 committed:
[MINOR] Various fixes of testcases (num jobs, ext functions)

mboehm7 committed:
[SYSTEMDS-11] Fix namespace handling of dml-bodied builtin functions

This patch fixes issues with using multiple dml-bodied builtin functions
in the same script. So far the default namespace was bound to the first
dml-bodied function file path. Now, we properly handle the file paths
for the default namespace separately.

corepointer committed:
[SYSTEMDS-112,122] Additional parfor image data augmentation tests

parfor test on image operations mirror, brightness using synthetic data

Closes #15.

mboehm7 committed:
[MINOR] Fix various tests (mlcontext, csv, seq, paramserv), refactoring

kev-inn committed:
[SYSTEMDS-113] New builtin functions for linear regression algorithms

This patch adds 4 new DML-bodied builtin functions:

  * lm
  * lmDS
  * lmCG
  * lmpredict

These functions are to be used for linear regression.
`lm` calls either `lmDS` or `lmCG` depending on the size of the input
matrices. `lmpredict` can be used
to predict the class of a feature vector with the weights usually
calculated before by use of `lm`.

This also adds a file used for documentation of builtin functions, which
currently only documents those new added builtins.

Beware that lm currently is non-functional and the test cases will fail,
because of a namespace conflict when using multiple different DML-bodied
builtin functions.

Closes #14.

mboehm7 committed:
[SYSTEMDS-11] Fix namespace handling of dml-bodied builtins, part 2

This patch now also fixes issues with dml-bodied builtin functions
calling other dml-bodied builtin functions.

mboehm7 committed:
[SYSTEMDS-18] Improved IPA passes for rewrites and constant propagation

With the goal of adding a hierarchy of higher-level abstractions of
dml-bodied functions, it is of utmost importance to properly collapase
these language abstractions and compile efficient plans, even during
initial compilation (not just relying on dynamic recompilation).

This patch makes additional improvements to our inter-procedural
analysis (IPA) framework, by
1) applying both static and dynamic rewrites in each IPA iteration
2) better constant propagation of literals across functions

On the new lm builtin function -- where the previous IPA initially
compiled 702 CP and 122 Spark instructions, but 0 Spark instructions
executed -- the changes (1) and (1+2) reduced the generated instructions
after rewrites to 159/52 and 31/2 respectively. Besides avoiding
unnecessary recompilation overhead this also makes debugging and
advanced optimizers much easier.

mboehm7 commented on docs/Tasks.txt:
```diff
@@ -71,7 +71,8 @@ SYSTEMDS-100 Various Fixes
  * 111 Time builtin function for script-level measurements            OK
  * 112 Image data augmentation builtin functions                      OK
- 
+ * 113 Builtin functions for linear regression algorithms             OK
```
Please, try to avoid leaking files across PRs.

kev-inn replied:
I will watch out for that in the future.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlock.java:
```diff
@@ -59,7 +59,23 @@ protected DenseBlock(int[] dims) {
+	    if (i == _odims.length) {
+	    	return _odims[i - 1];
+		}
+	    return _odims[i - 1] / _odims[i];
```
1) The formatting tabs vs spaces is off
2) Maybe rewrite to single return and nested ternary operators (to improve readability)

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlockBool.java:
```diff
@@ -155,26 +155,31 @@ public DenseBlock set(int r, int c, double v) {
 	public DenseBlock set(DenseBlock db) {
-		System.arraycopy(db.valuesAt(0), 0, _data, 0, _rlen*_odims[0]);
+		double[] data = db.valuesAt(0);
+	    for (int i = 0; i < _rlen*_odims[0]; i++) {
```
The formatting is off - please make a full pass over all modified classes in that regard as I'm not commenting on the remaining ones.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlockBool.java:
```diff
@@ -155,26 +155,31 @@ public DenseBlock set(int r, int c, double v) {
 	
 	@Override
 	public DenseBlock set(DenseBlock db) {
-		System.arraycopy(db.valuesAt(0), 0, _data, 0, _rlen*_odims[0]);
```
I would recommend to run a few experiments (maybe in a separate PR for performance improvements) and see if we could can something if the input DenseBlock is a DenseBlockBool (i.e., _data.clear(); _data.or(that._data))

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlockFP32.java:
```diff
@@ -149,7 +149,7 @@ public DenseBlock set(int r, int c, double v) {
 	@Override
 	public DenseBlock set(DenseBlock db) {
-		System.arraycopy(db.valuesAt(0), 0, _data, 0, _rlen*_odims[0]);
+		System.arraycopy(DataConverter.toFloat(db.valuesAt(0)), 0, _data, 0, _rlen*_odims[0]);
```
Again maybe in a different PR (perf), I'd like to avoid unnecessary array copies (toFloat()) just to use an array copy after that. Probably the best would be to have a stream/loop to set all elements by using get double and a special case for common input/output arrays.

kev-inn replied:
I added the Todos for performance, I guess this resolves this for now. Same for the DenseBlockBool

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlockLBool.java:
```diff
@@ -0,0 +1,366 @@
+	}
+
+	@Override
+	public void reset(int rlen, int[] odims, double v) {
```
It would be nice if we could reduce the code size of this method.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlockLBool.java:
```diff
@@ -0,0 +1,366 @@
+/*
+ * Modifications Copyright 2018 Graz University of Technology
```
Generally please always use the year of the first modification.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlockLBool.java:
```diff
@@ -0,0 +1,366 @@
+	public long countNonZeros() {
+		long nnz = 0;
+		for (int i = 0; i < _blocks.length - 1; i++) {
+			nnz += UtilFunctions.computeNnz(_blocks[i], 0, blockSize() * _odims[0]);
```
if we ensure that all bits past the logical size of a block are zero, we could simply call Bitset.cardinality.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlockLBool.java:
```diff
@@ -0,0 +1,366 @@
+		// that one DenseBlock uses a block to represent x rows the other one can actually fit x+y rows in a block.
+		while (true) {
+			int length;
+			if (srcPos != 0) {
```
Is this really necessary - except for the boolean DenseBlock all of these dense blocks should be aligned as we systematically compute the blocksize. So maybe its a good idea to use the same blocksize for boolean blocks as well because several binary operations would have the same problem.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlockLBool.java:
```diff
@@ -0,0 +1,366 @@
+
+	@Override
+	public DenseBlock set(int rl, int ru, int cl, int cu, DenseBlock db) {
+		boolean allColumns = cl == 0 && cu == _odims[0];
```
This allColumns variable is not used at all.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlockLFP32.java:
```diff
@@ -0,0 +1,331 @@
+		return _blocks.length == 1;
+	}
+
+	private boolean isReusable(int rlen, int[] odims) {
```
While thinking about this notion of reusable, I remembered why before we only checked for exact matches: if we reuse too large blocks, the memory requirements of operations change and we might run out of memory although we explicitly scheduled an operation in memory. Accordingly, let's start with a more conservative approach to reusing. This might also make it much simpler such that we can pull it into the DenseBlock implementation.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlockLInt32.java:
```diff
@@ -0,0 +1,332 @@
+			// More memory is needed
+		    int newBlockSize = Integer.MAX_VALUE / odims[0];
+		    int restBlockSize = rlen % newBlockSize;
+		    int newNumBlocks = (rlen / newBlockSize) + (restBlockSize == 0 ? 0 : 1);
```
Let's think about what parts of these implementations are almost identical can can be moved to the base classes.

mboehm7 commented on src/test/java/org/tugraz/sysds/test/component/tensor/DenseBlockSetRowTest.java:
```diff
@@ -0,0 +1,262 @@
+import org.junit.Assert;
+import org.junit.Test;
+import org.tugraz.sysds.common.Types.ValueType;
+import org.tugraz.sysds.runtime.data.*;
```
Please avoid wildcard imports as it could make it hard to find the exact classes (especially if there are multiple wildcards and classes with the same name in different packages)

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/util/UtilFunctions.java:
```diff
@@ -621,7 +614,14 @@ public static int computeNnz(BitSet a, int ai, int len) {
+	public static int computeNnz(String[] a, int ai, int len) {
+		int lnnz = 0;
+		for( int k=ai; k<ai+len; k++ )
+			lnnz += (a[k] != null && !a[k].isEmpty()) ? 1 : 0;
```
please also treat all Strings of value zero (0, 0.0, -0, +0, etc) as zero value.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/util/DataConverter.java:
```diff
@@ -51,6 +51,8 @@
 import org.tugraz.sysds.runtime.matrix.data.OutputInfo;
 import org.tugraz.sysds.runtime.meta.MatrixCharacteristics;
 
+import javax.rmi.CORBA.Util;
```
Unused import.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlockFP64.java:
```diff
@@ -187,4 +193,9 @@ public double get(int r, int c) {
 	}
+
+	@Override
+	public String getString(int[] ix) {
```
Apart from the String modifications, I guess we need to slightly generalize this implementation because otherwise it would not work for combinations of different value types (e.g., on set dense block). However, please be cautious as this double dense block is already in active use, so we need to make sure there are no performance regressions.

mboehm7 reviewed (commented):
Thanks @kev-inn for your hard work on this patch, which is a lot given the missing template support. Please address the comments, try to simplify the logic of reusing/allocation, and try to create abstractions for duplicated code wherever possible. Performance improvements can be postponed to another PR but please add related `TODO perf` comments.

mboehm7 commented:
In addition to the inline comments, please make a pass over the code and remove warnings where possible (my IDE shows 48 additional warnings)

mboehm7 committed:
[SYSTEMDS-18] Improved IPA and size inference w/ ops on empty matrices

mboehm7 committed:
[MINOR] Fix various tests (cumagg, ipa, indexing, rand, mmchain)

kev-inn committed:
Resolve syntactical problems

kev-inn committed:
Resolve further syntactical warnings

kev-inn committed:
Resolve some duplicate code

For now not all duplicate code was eliminated, because of performance concerns.

kev-inn committed:
Move isReusable to base class

kev-inn committed:
Change license dates and add some Todos

kev-inn committed:
Refactor LargeDenseBlocks to reduce code reuse

Might have impacts on performance which is why I skipped the DenseRowBlocks (not large) for now. Will refactor all except the FP64 soon, so we don't have performance regressions.

mboehm7 commented:
hmm, "to reduce code reuse" ;-) let me know once you're happy with the PR and I'll have another look

kev-inn commented:
> hmm, "to reduce code reuse" ;-) let me know once you're happy with the PR and I'll have another look

my bad, of course it should say "to improve code reuse". The next commit should be the last one.

kev-inn committed:
Improve code reuse for Dense Row Blocks

Remove duplicate code from DenseRowBlocks, but keeping the implementations for FP64 for now.

kev-inn commented:
@mboehm7 I think all issues you addressed should be resolved now. I have changed quite a bit to make the code easier to work with. Performance could be impacted, but not by much, if we consider the performance mistakes (iterating twice over array for copying, which was not necessary) it should actually have improved.

mboehm7 commented:
LGTM - thanks @kev-inn for the patch. I only made some minor cleanups: formatting (tabs vs spaces, ternary operators, TODO instead of ToDo), replaced the used `NotImplementedException` with its counterpart in `apache.commons.lang`, and added some additional TODOs.
